# Exploring Ebay Car Sales Data 

## Introduction:
Dataset of used cars from german ebay website.
## Goal:
To clean the data and analyse the included car listings.

## Data Dictionary

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

## Data Exploration

In [35]:
import pandas as pd
import numpy as np

In [36]:
## autos = pd.read_csv('autos.csv',encoding='UTF-8')
autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [37]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [38]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

*This information about autos dataset shows there are almost 6000 null observations in Vehicle type and 4500 null observations in fuel type.*

In [39]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

*Based on the first five rows, we can see that there are multiple places where data cleaning is required. At first glance:*
* Name - Name of the car has multiple underscrores with real names of the car hidden in the large string
* Seller - Privat instead of Private
* Vehicle Type - Error in description
* Odometer - Having km at the end of the kilometers driven makes it string. Needs to be numeric.


* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [40]:
## To check existing column names
autos_cols = autos.columns
autos_cols

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [41]:
autos.rename(columns= {"dateCrawled": "date_crawled", "offerType" : "offer_type","vehicleType": "vehicle_type",
                       "yearOfRegistration": "registration_year", "powerPS": "power_ps", 
                       "monthOfRegistration": "registration_month","notRepairedDamage": "unrepaired_damage",
                        "dateCreated": "ad_created","nrOfPictures": "nr_of_pictures","postalCode": "postal_code",
                        "lastSeen": "last_seen"
                          },inplace=True)

In [42]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuelType  \
0    manuell       158  andere  150,000km                   3      lpg   
1  automatik       286     7er  150,000km                   6   benzin   
2    manuell       102    golf   70,000km                   7   benzin   
3  automatik        71  fortwo   70,000km                   6   benzin   
4    manuell         0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

*As can be seen in the table above, the columns names have been changed from camelcase to snakecase for its ease of use and it shows exactly what the column represents. For example, its difficult to read yearOfRegistration, it makes it so much more easier, when we see registration_year as the column.*

*Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. - Examples of numeric data stored as text which can be cleaned and converted.*

In [43]:
## Descriptive statistics for all columns
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

*The number of pictures seems to have only one value 0. This can be dropped. Postal code is stored as numeric but should rather be stored as text. Price column is stored as text, which needs to be stored as Numeric. Price needs to be cleaned by removing the $ sign as prefix of the price and comma sign as well needs to be removed. Odometer column is also a numeric format which is stored as text. It has km as the suffix. The odometer column needs to be converted into text format.*

In [44]:
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [45]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [46]:
autos["postal_code"]=autos["postal_code"].astype(str)

In [47]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int32
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null int32
registration_month    50000 non-null int64
fuelType              45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null object
last_seen             50000 non-null object
dtypes: int32(2),

In [48]:
autos.rename(columns= {"odometer": "odometer_km"
                          },inplace=True)

In [49]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuelType  \
0    manuell       158  andere       150000                   3      lpg   
1  automatik       286     7er       150000                   6   benzin   
2    manuell       102    golf        70000                   7   benzin   
3  automatik        71  fortwo        70000                   6   benzin   
4    manuell         0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

  postal_code            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

*Now, following changes have been made to the table:*
* Price has been changed from text to numeric
* Odometer has been changed from text to numeric and the column name has changed from odometer to odometer_km
* Postal code has been changed from numeric to text


*There are a number of text columns where almost all the values are same. For example seller or offer_type. We will also analyse Price and Odometer_km column to check the minimum and maximum values and the outliers.*

In [50]:
##checking unique values in Price
autos["price"].unique().shape

(2357,)

In [51]:
autos["odometer_km"].unique().shape

(13,)

In [52]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [53]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [54]:
autos['price'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64

In [55]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [56]:
##Remove outliers from price column
## Syntax ----> df[df["col"].between(x,y)]
autos_v1 = autos[autos["price"].between(1,2732222)]

In [57]:
autos_v1.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuelType  \
0    manuell       158  andere       150000                   3      lpg   
1  automatik       286     7er       150000                   6   benzin   
2    manuell       102    golf        70000                   7   benzin   
3  automatik        71  fortwo        70000                   6   benzin   
4    manuell         0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

  postal_code            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [58]:
autos_v1.describe()

price  registration_year      power_ps    odometer_km  \
count  4.857000e+04       48570.000000  48570.000000   48570.000000   
mean   6.002279e+03        2004.754231    117.194812  125770.022648   
std    1.444526e+04          88.639460    200.640396   39788.530021   
min    1.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.200000e+03        1999.000000     71.000000  125000.000000   
50%    3.000000e+03        2004.000000    107.000000  150000.000000   
75%    7.490000e+03        2008.000000    150.000000  150000.000000   
max    1.300000e+06        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures  
count        48570.000000         48570.0  
mean             5.782294             0.0  
std              3.685734             0.0  
min              0.000000             0.0  
25%              3.000000             0.0  
50%              6.000000             0.0  
75%              9.000000             0.0  
max             12.000000             0.0

*So we removed 0 and 99999999 from the price column as these seemed like an outlier.*

## Understanding the date distribution

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively.

The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [59]:
## Checking the data
autos_v1[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [66]:
# Select the date crawled series. Take first 10 characters in the datecrawled. Then sort it by date.
# To include missing values in the distribution and to use percentages instead of counts,
# chain the Series.value_counts(normalize=True, dropna=False) method.


autos_v1["date_crawled"].str[0:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025324
2016-03-06    0.014042
2016-03-07    0.036010
2016-03-08    0.033292
2016-03-09    0.033086
2016-03-10    0.032180
2016-03-11    0.032572
2016-03-12    0.036916
2016-03-13    0.015668
2016-03-14    0.036545
2016-03-15    0.034280
2016-03-16    0.029607
2016-03-17    0.031645
2016-03-18    0.012909
2016-03-19    0.034775
2016-03-20    0.037883
2016-03-21    0.037389
2016-03-22    0.033004
2016-03-23    0.032222
2016-03-24    0.029339
2016-03-25    0.031604
2016-03-26    0.032201
2016-03-27    0.031089
2016-03-28    0.034857
2016-03-29    0.034116
2016-03-30    0.033683
2016-03-31    0.031830
2016-04-01    0.033683
2016-04-02    0.035475
2016-04-03    0.038604
2016-04-04    0.036504
2016-04-05    0.013095
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


*We can see that there is uniformity in the distribution of the listings crawled daily.*

In [67]:
autos_v1["ad_created"].str[0:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033148
2016-03-10    0.031892
2016-03-11    0.032901
2016-03-12    0.036751
2016-03-13    0.017006
2016-03-14    0.035186
2016-03-15    0.034013
2016-03-16    0.030121
2016-03-17    0.031295
2016-03-18    0.013589
2016-03-19    0.033683
2016-03-20    0.037945
2016-03-21 

*We can see the majority of the ads created was within a similar time frame as the crawling dates. The lower frequency of ads created prior to these dates are understandable as once the sales have been completed, the ads can be delisted and won't appear on the website prior to the dates recorded by the crawler*

In [68]:
autos_v1["last_seen"].str[0:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005394
2016-03-08    0.007412
2016-03-09    0.009594
2016-03-10    0.010665
2016-03-11    0.012374
2016-03-12    0.023780
2016-03-13    0.008894
2016-03-14    0.012600
2016-03-15    0.015874
2016-03-16    0.016450
2016-03-17    0.028083
2016-03-18    0.007350
2016-03-19    0.015833
2016-03-20    0.020651
2016-03-21    0.020630
2016-03-22    0.021371
2016-03-23    0.018530
2016-03-24    0.019765
2016-03-25    0.019209
2016-03-26    0.016800
2016-03-27    0.015648
2016-03-28    0.020877
2016-03-29    0.022359
2016-03-30    0.024768
2016-03-31    0.023780
2016-04-01    0.022792
2016-04-02    0.024933
2016-04-03    0.025201
2016-04-04    0.024480
2016-04-05    0.124768
2016-04-06    0.221804
2016-04-07    0.131933
Name: last_seen, dtype: float64


*There is a spike seen in the distribution for the last three days for the 'last_seen' column. This last_seen column tells us of when an ad has been delisted from the site, most likely due to a sale. This high distribution observed during the last 3 days of the crawler recording is most likely not due to a drastic increase in sales as the days prior to these had a roughly similar distribution.*

In [69]:
## Registration year distribution
autos_v1["registration_year"].describe()

count    48570.000000
mean      2004.754231
std         88.639460
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

*As can be seen the minimum value of registration year is 1000 and maximum value of registration year is 9999. Now both of these are outliers.*

*One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:*

*The minimum value is 1000, before cars were invented
The maximum value is 9999, many years into the future
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s. First car ever made was 1893. So we can safely assume to have the registration year of 1900.*

In [72]:
autos_v2 = autos_v1[autos_v1["registration_year"].between(1900,2016)]

In [73]:
autos_v2.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuelType  \
0    manuell       158  andere       150000                   3      lpg   
1  automatik       286     7er       150000                   6   benzin   
2    manuell       102    golf        70000                   7   benzin   
3  automatik        71  fortwo        70000                   6   benzin   
4    manuell         0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

  postal_code            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [75]:
autos_v2["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007261
1992    0.007947
1993    0.009103
1994    0.013473
1995    0.026282
1996    0.029409
1997    0.041790
1998    0.050615
1999    0.062074
2000    0.067601
2001    0.056462
2002    0.053249
2003    0.057812
2004    0.057897
2005    0.062888
2006    0.057191
2007    0.048773
2008    0.047445
2009    0.044681
2010    0.034036
2011    0.034764
2012    0.028060
2013    0.017200
2014    0.0142

*Majority of the cars were registered in the year starting 1997 till 2012. There has been a visible decline in the car registration since 2012. May be the number of cars being sold are less compared to early 2000s.*

## Exploring car sales distribution by brands

In [76]:
autos_v2["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [81]:
brand_count=autos_v2["brand"].value_counts(normalize=True)
brand_count

volkswagen        0.211284
bmw               0.110054
opel              0.107570
mercedes_benz     0.096453
audi              0.086557
ford              0.069914
renault           0.047145
peugeot           0.029838
fiat              0.025639
seat              0.018271
skoda             0.016407
nissan            0.015272
mazda             0.015187
smart             0.014158
citroen           0.014008
toyota            0.012702
hyundai           0.010024
sonstige_autos    0.009832
volvo             0.009146
mini              0.008761
mitsubishi        0.008225
honda             0.007840
kia               0.007069
alfa_romeo        0.006640
porsche           0.006126
suzuki            0.005933
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001499
trabant           0.001392
r

*Based on the brand aggregation, we can select the top 6 brands
* volkswagen        0.211284
* bmw               0.110054
* opel              0.107570
* mercedes_benz     0.096453
* audi              0.086557
* ford              0.069914

*They account for almost 68-70% of the total car sales.*

In [105]:
brands = autos_v2['brand'].value_counts(normalize = True)
top_brands = brands[brands > 0.06].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [106]:
brand_mean_price={}
for key in top_brands:
    brand_only=autos_v2[autos_v2["brand"]==key]
    mean_price=brand_only["price"].mean() 
    brand_mean_price[key]=int(mean_price)

In [107]:
brand_mean_price

{'volkswagen': 5604,
 'bmw': 8571,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 4054}

*We can see that there is a price gap between the top 5 brands in the sales data. We can see that cars manufactured by Audi, BMW and Mercedes Benz tend to be priced higher than the competition. Opel is the least expensive of the top 5 brands while Volkswagen is in between. This could be one of the reasons for the popularity of Volkswagen cars.*

* We observed that in the top 6 brands, there's a distinct price gap.

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between


## Mileage Comparision between top 6 brands
*For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:
* It's difficult to compare more than two aggregate series objects if we want to extend to more columns
* We can't compare more than a few rows from each series object
* We can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

In [108]:
brand_mean_mileage={}
for key in top_brands:
    brand_only=autos_v2[autos_v2["brand"]==key]
    mean_mileage=brand_only["odometer_km"].mean() 
    brand_mean_mileage[key]=int(mean_mileage)

In [109]:
brand_mean_mileage

{'volkswagen': 128711,
 'bmw': 132575,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [112]:
bmp_series=pd.Series(brand_mean_price)
bmm_series=pd.Series(brand_mean_mileage)

In [113]:
bmp_series

volkswagen       5604
bmw              8571
opel             2975
mercedes_benz    8628
audi             9336
ford             4054
dtype: int64

In [114]:
bmm_series

volkswagen       128711
bmw              132575
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64

In [117]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df['mean_mileage']=bmm_series
df

mean_price  mean_mileage
volkswagen           5604        128711
bmw                  8571        132575
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 4054        124266

*We can see that the car mileage doesn't vary as much as the prices. We can see that the more expensive brands generally tend to have higher mileage with the only outlier being Opel.*

In [119]:
autos_v2["model"].value_counts().head(1)
autos_v2.loc[autos["model"]=='golf',"brand"].head(1)

2    volkswagen
Name: brand, dtype: object

*The most common car model is Volkswagen Golf.*

## Conclusion

*After cleaning the ebay car sales data, we found out that the top 6 brands account for 70% of the sales. The most common car model is Volkswagen Golf. We also saw that car mileage doesnt vary much as the prices.*